fork-runner/src/main/java/com/shazam/fork/model/PoolTestCaseFailureAccumulator.java

```diff
@@ -1,69 +1,78 @@
+/*
+ * Copyright 2018 Shazam Entertainment Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
 package com.shazam.fork.model;
 
-import com.google.common.base.Predicate;
-import com.google.common.collect.HashMultimap;
-import com.google.common.collect.ImmutableList;
 import com.google.common.collect.SetMultimap;
 
-import static com.google.common.collect.FluentIterable.from;
+import java.util.function.Predicate;
+
+import static com.google.common.collect.HashMultimap.create;
+import static com.google.common.collect.Multimaps.synchronizedSetMultimap;
 
 /**
  * Class that keeps track of the number of times each testCase is executed for device.
  */
 public class PoolTestCaseFailureAccumulator implements PoolTestCaseAccumulator {
-
-    private SetMultimap<Pool, TestCaseEventCounter> map = HashMultimap.<Pool, TestCaseEventCounter>create();
+    private final SetMultimap<Pool, TestCaseEventCounter> testCaseCounters = synchronizedSetMultimap(create());
 
     @Override
     public void record(Pool pool, TestCaseEvent testCaseEvent) {
-        if (!map.containsKey(pool)) {
-            map.put(pool, createNew(testCaseEvent));
+        if (!testCaseCounters.containsKey(pool)) {
+            testCaseCounters.put(pool, createNew(testCaseEvent));
```

Comment:
Shouldn't this go inside a synchronized block?

Reply:
No if I understand correctly. We have to wrap it inside a synchronized block only when we iterate over this collection. More info here https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/Multimaps.html#synchronizedMultimap(com.google.common.collect.Multimap)